rust/agents/scraper/src/settings.rs

```diff
@@ -1,13 +1,80 @@
+use eyre::WrapErr;
 use std::collections::HashMap;
+use std::env;
 
-use abacus_base::decl_settings;
-use abacus_base::{ChainSetup, IndexSettings, OutboxAddresses};
-
-// TODO: Make it so the inherited settings better communicate that the `outbox`
-// config is not needed for the scraper.
-decl_settings!(Scraper {
-    /// Configurations for contracts on the outbox chains
-    outboxes: HashMap<String, ChainSetup<OutboxAddresses>>,
-    /// Index settings by chain
-    indexes: HashMap<String, IndexSettings>,
-});
+use abacus_base::load_settings_object;
+use abacus_base::{AgentSettings, ApplicationSettings, ChainSettings};
+
+/// Scraper settings work a bit differently than other agents because we need to
+/// load the information for all of the chains.
+///
+/// The same basic principals apply as to what files/envs are read from and
+/// their order of precedence. You can read more in the `load_settings_object`
+/// docs.
+///
+/// You will need to define `RUN_DOMAINS` which is a comma seperated list of
+/// chain names. Then for each domain if a config file is to be read from
+/// `BASE_CONFIG_${DOMAIN}` should be specified. The env vars will read with the
+/// prefix `HYP_SCRAPER_${DOMAIN}_*` for chain configs, and just `HYP_SCRAPER_*`
+/// for any generic application configs.
+///
+/// `HYP_BASE_*` will still be a default for all chains, so you could define
+/// `HYP_BASE_INDEX_CHUNK` and define a default chunk size for all chains.
+///
+/// You still need the `BASE_CONFIG` if you want to have any application
+/// settings loaded via a config file.
+#[derive(Debug)]
+pub struct ScraperSettings {
+    /// settings by domain id for each domain
+    pub chains: HashMap<u32, ChainSettings>,
+    pub app: ApplicationSettings,
+}
+
+impl AsRef<ApplicationSettings> for ScraperSettings {
+    fn as_ref(&self) -> &ApplicationSettings {
+        &self.app
+    }
+}
+
+impl AgentSettings for ScraperSettings {
+    type Error = eyre::Report;
+
+    fn new() -> Result<Self, Self::Error> {
+        let app = load_settings_object::<_, &str>(
+            "scraper",
+            env::var("BASE_CONFIG").ok().as_deref(),
+            &[],
+        )
+        .context("Loading application settings")?;
+        let uppercase_chain_list = env::var("RUN_DOMAINS")
+            .expect("Must specify run domains for scraper")
+            .to_ascii_uppercase();
+        let chains: Vec<&str> = uppercase_chain_list.split(',').collect();
+
+        let chain_settings = chains
+            .iter()
+            .map(|chain_name| {
+                let config_file_name = env::var(&format!("BASE_CONFIG_{chain_name}"))
+                    .expect("Must specify config file for all domains in $RUN_DOMAINS");
+                let ignore_prefixes = [
```

Comment:
Because it will cause a panic if it sees a config for the inbox that is on the same chain as the outbox. Which is because fundamentally that does not make sense, and when we are parsing configs like this we have a bunch of things with slightly overlapping views.

Suffice it to say this _works_ but it would be nice to have a better solution for config parsing.

Reply:
Done